Framework/Algorithms/src/FindPeaks.cpp

```diff
@@ -632,24 +637,52 @@ void FindPeaks::findPeaksUsingMariscotti() {
 API::MatrixWorkspace_sptr FindPeaks::calculateSecondDifference(
     const API::MatrixWorkspace_const_sptr &input) {
   // We need a new workspace the same size as the input ont
-  MatrixWorkspace_sptr diffed = WorkspaceFactory::Instance().create(input);
+  MatrixWorkspace_sptr diffed = 0;
```

Comment:
I think if you use the factory functions I provide in `DataObjects/WorkspaceCreation.h` like this:
```cpp
auto diffed = create<MatrixWorkspace>(input);
```
it will automatically copy all X arrays for you and you can completely remove all the code duplication you added below (unless I am overlooking something), and remove the line
```cpp
    // Copy over the X values
    diffed->setSharedX(i, input->sharedX(i));
```
below.

The only actual thing that needs to be done is adapt those factory methods to work in a way that permits variable X lengths, namely use `setHistogram` instead of `setSharedX` here: https://github.com/mantidproject/mantid/blob/master/Framework/DataObjects/inc/MantidDataObjects/WorkspaceCreation.h#L267 (detail to figure out: how to set the correct Y/E values, let me know if you have trouble with that).

Reply:
Simon, Thanks.  I will refactor the code according to your suggestion.

Reply:
I had the same thoughts here too. As it wastes resources to produce a multi-spectrum MatrixWorkspace if there is only one spectrum to find peaks at, I create a single-spectrum MatrixWorkspace other than a multip-spectrum workspace.

Reply:
I tried to use 
```auto diffed = create<MatrixWorkspace>(input)```
but I got compiling error as
```
/Users/wzz/Mantid/Framework/Algorithms/src/FindPeaks.cpp:659:14: error: no matching function for call to 'create'
    diffed = Mantid::DataObjects::create<API::MatrixWorkspace>(input);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/wzz/Mantid/Framework/DataObjects/inc/MantidDataObjects/WorkspaceCreation.h:260:35: note: candidate template ignored: disabled by 'enable_if' [with T = Mantid::API::MatrixWorkspace, P = boost::shared_ptr<Mantid::API::MatrixWorkspace>]
          typename std::enable_if<std::is_base_of<API::MatrixWorkspace,
```

On the other hand, so far, in master, this is what I see for method create<>(parent):
```
template <class T, class P,
          typename std::enable_if<std::is_base_of<API::MatrixWorkspace,
                                                  P>::value>::type * = nullptr>
std::unique_ptr<T> create(const P &parent) {
  const auto numHistograms = parent.getNumberHistograms();
  auto ws =
      create<T>(parent, numHistograms, detail::stripData(parent.histogram(0)));
  for (size_t i = 0; i < numHistograms; ++i) {
    ws->setSharedX(i, parent.sharedX(i));
  }
  return ws;
}
```

Reply:
@wdzhou Yes, the methods uses a reference to the parent, not a `shared_ptr`.